src/Raven.Server/Config/JsonConfigFileModifier.cs

```diff
@@ -0,0 +1,225 @@
+
+        var fileName = Path.GetFileName(_path);
+        return context.ReadObject(jsonConfig, fileName);
+    }
+
+    private void FillJsonFromFile(DynamicJsonValue jsonConfig)
+    {
+        try
+        {
+            using var fs = OpenFile(_path, FileMode.Open, FileAccess.ReadWrite);
+            var configBuilder = new ConfigurationBuilder();
+            configBuilder.AddJsonStream(fs);
+            var config = configBuilder.Build();
+            var orderConfig = config.AsEnumerable().OrderBy(x => x.Key);
+            var dicConfig = new Dictionary<string, object>();
+            foreach (var (key, value) in orderConfig)
+            {
+                if (value == null)
+                    continue;
+                
+                var directKey = key.Replace(':', '.');
+                if (dicConfig.TryAdd(directKey, value) == false)
+                    continue;
+
+                jsonConfig[directKey] = value;
+            }
+        }
+        catch (FileNotFoundException)
+        {
+        }
+    }
+
+    private static FileStream OpenFile(string path, FileMode fileMode, FileAccess fileAccess)
+    {
+        try
+        {
+            return SafeFileStream.Create(path, fileMode, fileAccess);
+        }
+        catch (Exception e) when (e is UnauthorizedAccessException or SecurityException)
+        {
+            throw new UnsuccessfulFileAccessException(e, path, fileAccess);
+        }
+    }
+
+    private async Task PersistConfiguration(BlittableJsonReaderObject json)
+    {
+        using var tempFile = new TempFile(_path);
+
+        await using (var file = OpenFile(tempFile.Path, FileMode.Create, FileAccess.ReadWrite))
+        await using (var streamWriter = new StreamWriter(file))
+        await using (var writer = new JsonTextWriter(streamWriter))
+        await using (var reader = new BlittableJsonReader())
+        {
+            writer.Formatting = Formatting.Indented;
+            reader.Initialize(json);
+
+            await writer.WriteTokenAsync(reader);
+                
+            await writer.FlushAsync();
+            await streamWriter.FlushAsync();
+            file.Flush(true);
+        }
+
+        Validate(tempFile.Path);
+        SwitchTempWithOriginalAndCreateBackup(tempFile.Path);
+    }
+
+    private void SwitchTempWithOriginalAndCreateBackup(string tempPath)
+    {
+        try
+        {
+            if (File.Exists(_path))
+            {
+                File.Replace(tempPath, _path, _path + ".bak");
+            }
+            else
+            {
+                File.Move(tempPath, _path);
+            }
+            
+            if (PlatformDetails.RunningOnPosix)
+                Syscall.FsyncDirectoryFor(_path);
+        }
+        catch (Exception e) when (e is UnauthorizedAccessException or SecurityException)
+        {
+            throw new UnsuccessfulFileAccessException(e, _path, FileAccess.Write);
+        }
+    }
+
+    private class TempFile : IDisposable
```

Comment:
this is used in one place? overkill IMO, also IIRC DeleteOnClose was fixed in Linux?

Reply:
done

Reply:
`DeleteOnClose` is not useful here since I need the file to be moved or replaced after I close the stream.
Anyway, I got rid of the `TempFile` class.